README.md
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]



<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/wouwehb/AR-MultiPlatformTool">
    <img src="data/img/logo.png" alt="Logo" width="80" height="80">
  </a>

  <h3 align="center">AR Multiplatform Tool</h3>

  <p align="center">
    A tool to help you analyze metadata of your Turned-based games.
    <br />
    <a href="https://github.com/wouwehb/AR-MultiPlatformTool"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/wouwehb/AR-MultiPlatformTool">View Demo</a>
    ·
    <a href="https://github.com/wouwehb/AR-MultiPlatformTool/issues">Report Bug</a>
    ·
    <a href="https://github.com/wouwehb/AR-MultiPlatformTool/issues">Request Feature</a>
  </p>
</p>



<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary><h2 style="display: inline-block">Table of Contents</h2></summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgements">Acknowledgements</a></li>
  </ol>
</details>



## About The Project

[<img src="data/img/rocket_background.png" alt="rocket_Background" width="400" height="275">](https://i.imgur.com/RF0TXKP.png)

 Automatically hook your game up to the api to automatically perform various checks and analyze post-game data. like notify you of people with unusualy high win rates; are the cheaters, is their build/play-style overpowered? This api is just a frame-work. Feel free to tinker with it to fulfill your every need.


## Getting Started

To get a local copy up and running follow these simple steps.


### Prerequisites

You will need the latest version of npm
* npm
  ```sh
  npm install npm@latest -g
  ```

You will also require your own database

### Installation

1. Clone the repo
   ```sh
   git clone https://github.com/wouwehb/AR-MultiPlatformTool
   ```
2. Install NPM packages
   ```sh
   npm install
   ```
3. Hook up the api to your game's back-end



## Usage




## Roadmap

See the [open issues](https://github.com/wouwehb/AR-MultiPlatformTool/issues) for a list of proposed features (and known issues).



## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request



## License

Distributed under the MIT License. See `LICENSE` for more information.



## Contact

Your Name - [@brandendeAap](https://twitter.com/BrandendeAap) - wouter.verthongen@student.ehb.be

Project Link: [https://github.com/wouwehb/AR-MultiplatformTool](https://github.com/wouwehb/AR-MultiPlatformTool)



## Acknowledgements

* [Tom Verheirstraeten](https://github.com/tomverheirstraeten/easywavespawner)
* [Jan Everaert](https://github.com/Everaertjan)





<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/wouwehb/AR-MultiplatformTool.svg?style=for-the-badge
[contributors-url]: https://github.com/wouwehb/AR-MultiplatformTool/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/wouwehb/AR-MultiplatformTool.svg?style=for-the-badge
[forks-url]: https://github.com/wouwehb/AR-MultiplatformTool/network/members
[stars-shield]: https://img.shields.io/github/stars/wouwehb/AR-MultiplatformTool.svg?style=for-the-badge
[stars-url]: https://github.com/wouwehb/AR-MultiplatformTool/stargazers
[issues-shield]: https://img.shields.io/github/issues/wouwehb/AR-MultiplatformTool.svg?style=for-the-badge
[issues-url]: https://github.com/wouwehb/AR-MultiplatformTool/issues
[license-shield]: https://img.shields.io/github/license/wouwehb/AR-MultiplatformTool.svg?style=for-the-badge
[license-url]: https://github.com/wouwehb/AR-MultiplatformTool/blob/main/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/wouwehb


CHANGELOG.md
## 0.0.1 - First Release
* Created base Skeletal structure
* Added simple DB
* Added example endpoints

api/__test__/server.test.js
const supertest = require("supertest");
const app = require("../src/server.js");
const request = supertest(app);

describe("Add a Game",()=>{

});

describe("get a game by ID", ()=>{
    
    test("between 0-1billion" , async (done) =>{
        try {
            const response = await request.get('/getGameById/-500');
            console.log("response:" + response);
            expect(response.status).toBe(400);
            done();
        } catch (error) {
            console.log(error);
        }
    });

    test("is an integer" , async (done) =>{
        try {
            const response = await request.get('/getGameById/:somethinginthewayshemovesme/')
            expect(response.status).toBe(400)
            done();
        } catch (error) {
            console.log(error)
        }
    });
});


describe("Get  win percentage  by Player",() =>{
    
    test("no funny business" , async (done) =>{
        try {
            const response = await request.get('/getWinPercentByPlayer/"DROP_TABLES();')
            expect(response.status).toBe(400)
            done();
        } catch (error) {
            console.log(error)
        }
    });

    test("not an  empty string" , async (done) =>{
        try {
            const response = await request.get('/getWinPercentByPlayer/" "')
            expect(response.status).toBe(400)
            done();
        } catch (error) {
            console.log(error)
        }
    });

    
});


describe("Get Players by minimum win percentage",() =>{
    //percent must between 0 and 100
    //percent is a number
    //input can't be empty
    //if return is empty give error message
    
});

describe("Get Games made during x time period",()=>{

    //time period must be correct format



});

describe("get games that ended after x seconds",()=>{
        test("is an integer" , async (done) =>{
        try {
            const response = await request.get('/getGamesByDuration/%%%"')
            expect(response.status).toBe(400)
            done();
        } catch (error) {
            console.log(error)
        }
    });


    test("value needs to be on db" , async (done) =>{
        try {
            const response = await request.get('/getGamesByDuration/900')
            expect(response.status).toBe(404)
            done();
        } catch (error) {
            console.log(error)
        }
    });

});


api/__test__/helpers.test.js
const Helpers = require("../src/utils/helpers.js");

describe("test test: sum", () => {
    test("Returns sum of the numbers", async () => {
        expect(Helpers.sum(1, 2)).toBe(3);
    });
});

/**
 * Unit Test
 */

describe("helper function; special character tester", () => {

    test("no ';'" , async (done) =>{
        try {
            expect(Helpers.specialCharacter(";")).toBe(true);
            done();
        } catch (error) {
            console.log(error);
        }
    });

    test("no ' " , async (done) =>{
        try {
            expect(Helpers.specialCharacter("'")).toBe(true);
            done();
        } catch (error) {
            console.log(error);
        }
    });

        
});

api/__test__/end_to_end.test.js
const supertest = require("supertest");
const app = require("../src/server.js");
const request = supertest(app);


/**
 * Testflow
 * ----------------------------------------------
 * Two players are added
 * A game has been played and is added to the Db
 * Developer checks for any irregularities
 * Developer recieves data
 * Developer bans player and deletes player info
 */

let player1;
let player2;

describe("add players", () => {
    test("add a player", async (done) => {
        try {
            const response = await request.post('/addPlayer/');
            expect(response.status).toBe(200);
            console.log("RESPONSE.BODY.uuid: " + response.body.uuid);
            player1 = response.body.uuid;

            done();
        } catch (error) {
            console.log(error);
        }
    });

    test("add a second player", async (done) => {
        try {
            const response = await request.post('/addPlayer/');
            expect(response.status).toBe(200);
            player2 = response.body.uuid;
            done();
        } catch (error) {
            console.log(error);
        }
    });
});

describe("let's play a game", () => {

    let gameData;
    let gameUuid;

    //A game has been played and is added to the Db

    test("add a game", async (done) => {
        try {
            gameData = {
                title: "Civilization XI Game 1234567890",
                rounds: 12,
                duration: 123,
                players: [player1, player2],
                winner: player1
            }

            const response = await request.post("/AddGame").send(gameData);
            expect(response.status).toBe(201);
            gameUuid = (response.body.uuid);
            done();
        } catch (error) {
            console.log(error)
        }
    })
    //check if game really is added
    test("check if game is on database", async (done) => {
        try {
            const response = await request.get(`/getGameByUUID/${gameUuid}`);
            expect(response.status).toBe(200);
            done();
        } catch (error) {
            console.log(error);
        }
    })

    //Developer checks for any irregularities

    test("check winrate player 1", async (done) => {
        try {
            const response = await request.get(`/getWinPercentByPlayer/${player1}`);
            expect(response.body.winrate).toBe(1);
            done();
        } catch (error) {
            console.log(error);
        }
    })


    //Developer deletes deletes player info

    test("delete player", async (done) => {

        try {
            const response = await request.delete(`/deletePlayer/${player1}`)
            expect(response.status).toBe(200);
            done()
        } catch (error) {
            console.log(error);
        }
    });

    //check if player really is deleted

    test('look for players between players', async (done) => {

        try {
            const response = await request.get(`/getPlayerById/${player1}`)
            expect(response.status).toBe(404);
            done();
        } catch (error) {
            console.log(error);
        }
    })

})

api/src/server.js
const express = require('express')
const bodyParser = require('body-parser');
const http = require('http');
const Helpers = require('./utils/helpers.js');
const cors = require("cors");
const {
  doesNotMatch
} = require('assert');
const {
  count
} = require('console');
const {
  isString
} = require('util');

const port = 3001


const pg = require('knex')({
  client: 'pg',
  version: '9.6',
  searchPath: ['knex', 'public'],
  connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/ARdb'
});


const app = express();
http.Server(app);
app.use(cors());

app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    extended: true
  })
);


app.get('/test', (req, res) => {

  res.status(200).send();
});

/**
 * returns html file with possible connections
 * @param: none
 * @returns: root.html
 */

app.get('/', async (req, res) => {

  res.sendFile(__dirname + '/root.html');

});

///////////////////////////////
//    Handle Incoming Data   //
///////////////////////////////

/**
 * Adds a game
 * @param: body with uuid, title, players, the winner, rounds and duration
 * @returns: uuid of game
 */

app.post('/addGame', async (req, res) => {

  console.log("Added a Game");

  const uuid = Helpers.generateUUID();

  const AddGame = await pg
    .table("games")
    .insert({
      uuid: uuid,
      title: req.body.title,
      rounds: req.body.rounds,
      duration: req.body.duration
    })
    .catch((error) => {
      console.log(error);
    });

  for (const player of req.body.players) {
    console.log(player);
    AddPlayedGame(uuid, player);
  }

  const AddWinner = await pg
    .table("played_games")
    .where({
      player_id: req.body.winner
    })
    .update({
      winner: true
    })
    .then(() => {
      console.log(`winner ${req.body.winner} has been updated`)
      res.status(201).send({
        uuid: uuid
      });
    })
    .catch((error) => {
      console.log(error);
    });
});

async function AddPlayedGame(uuid, player) {

  const AddPlayedGame = await pg
    .table("played_games")
    .where()
    .insert({
      game_id: uuid,
      player_id: player,
      winner: false
    }).catch((error) => {
      console.log(error);
    });
}

/**
 * add a player
 * @param: none
 * @returns: player uuid
 */

app.post('/addPlayer', async (req, res) => {

  const uuid = Helpers.generateUUID();

  const resultGames = await pg
    .table("players")
    .insert({
      uuid: uuid
    })
    .then(async function () {
      res.status(200).send({
        uuid: uuid
      });
    })
    .catch((error) => {
      console.log(error);
    });
});


/**
 * get all games
 * @param: none
 * @returns: json with all games
 */

app.get('/getAllGames', async (req, res) => {

  console.log("get all games");

  const result = await pg.select("*").from("games");
  res.json({
    res: result,
  });

});


/**
 * get a specific game by id
 * @param: id
 * @returns: game with id = param.body.id
 */

app.get('/getGameById/:id', async (req, res) => {

  let canPass = true;
  let id = parseInt(req.params.id);


  if (Helpers.specialCharacter(req.params.id)) {
    canPass = false
  }

  if (!Number.isInteger(id)) {
    canPass = false
  }

  if (id < 0 || id > 100000000000) {

    canPass = false
  }

  if (canPass) {

    try {
      const result = await pg
        .select("*")
        .from("games")
        .where({
          id: id
        })
      if (result.length < 1) {
        res.status(404).send();
      } else {
        res.send(result)
      }
    } catch (error) {
      res.send(error);
    }

  } else {
    res.sendStatus(400);
  }
});

/**
 * get a specific game by uuid
 * @param: uuid
 * @returns: game with uuid = param.body.uuid
 */

app.get('/getGameByUUID/:uuid', async (req, res) => {

  let canPass = true;


  if (Helpers.specialCharacter(req.params.uuid)) {
    canPass = false
  }
  if (canPass) {
    try {
      const result = await pg
        .select("*")
        .from("games")
        .where({
          uuid: req.params.uuid
        })
      if (result.length < 1) {
        res.status(404).send();
      } else {
        res.status(200).send(result);
      }
    } catch (error) {
      res.send(error);
    }

  } else {
    res.sendStatus(400);
  }
});

/**
 * get a specific player by uuid
 * @param: uuid
 * @returns: a player with uuid = param.body.uuid
 */

app.get('/getPlayerById/:uuid', async (req, res) => {

  if (Helpers.specialCharacter(req.params.uuid)) {
    res.sendStatus(400);
  } else {
    const result = await pg
      .select("*")
      .from("players")
      .where({
        uuid: req.params.uuid
      }).then(async (data) => {
        if (data.length >= 1) {
          res.json({
            res: data,
          });
        } else {
          res.status(404).send();
        }
      });
  }

});

/**
 * get all players
 * @param: none
 * @returns: json with all players on the database
 */

app.get('/getAllPlayers', async (req, res) => {

  const result = await pg.select("*").from("players");
  res.json({
    res: result,
  });

});

/**
 * delete a game by uuid
 * @param: uuid
 * @returns: status code
 */

app.delete('/deleteGame/:uuid', async (req, res) => {

  console.log("delete game");

  if (Helpers.specialCharacter(req.params.uuid)) {
    res.sendStatus(400);
  } else {
    const result = await pg
      .from("games")
      .where({
        uuid: req.params.uuid
      })
      .del().then((data) => {
        res.json(data);
      })
      .catch(() => res.status(404).send());
  }

});

/**
 * delete a player by uuid
 * @param: uuid
 * @returns: status code
 */

app.delete('/deletePlayer/:uuid', async (req, res) => {

  console.log("delete game");

  if (Helpers.specialCharacter(req.params.uuid)) {
    res.sendStatus(400);
  } else {
    const result = await pg
      .from("players")
      .where({
        uuid: req.params.uuid
      })
      .del().then((data) => {
        res.json(data);
      })
      .catch(() => res.status(404).send());
  }

});


///////////////////////////////
//        Analyze Data       //
///////////////////////////////


/**
 * calculate win percent by player
 * @param: uuid
 * @returns: float with value between 0-10
 */

app.get('/getWinPercentByPlayer/:uuid', async (req, res) => {

  if (Helpers.specialCharacter(req.params.uuid) || req.params.uuid === null || req.params.uuid === " " || typeof req.params.uuid !== "string") {
    res.sendStatus(400);

  } else {
    let totalGames;
    let WonGames;
    let result;

    const totalGamesResult = await pg
      .select("player_id")
      .from('played_games')
      .where({
        player_id: req.params.uuid
      }).then(async (data) => {
        totalGames = data.length
      })

      .catch((error) => {
        console.log(error);
      });


    const totalWonGamesResult = await pg
      .select("player_id")
      .from('played_games')
      .where({
        player_id: req.params.uuid,
        winner: true
      }).then(async (data) => {
        WonGames = data.length
        if (totalGames != 0 || wonGames === 0) {
          result = WonGames / totalGames;

        } else {
          result = 0;
        }

        if (result != null) {
          res.json({
            winrate: result,
          });
        } else {
          res.json({
            winrate: 0,
          });
        }


      })

      .catch((error) => {
        console.log(error);
      });
  }


});

/**
 * get games with a duration lower than x in seconds
 * @param: duration 
 * @returns: list of games who fullfill the criterea.
 */

//get all games with duration below x min
app.get('/getGamesByDuration/:duration', async (req, res) => {

  let duration = parseInt(req.params.duration);

  console.log("duration: " + duration);

  if (isNaN(duration) || duration === null || !Number.isInteger(duration)) {
    res.status(404).send();;
  } else {
    const result = await pg
      .select("*")
      .from("games")
      .where({
        duration: duration
      }).then(async (data) => {
        if (data.length >= 1) {
          res.json({
            res: data,
          });
        } else {
          res.status(404).send();
        }
      }).catch(() => res.status(400).send());
  }

});

///////////////////////////////
//       Create Tables       //
///////////////////////////////


async function initialiseTables() {
  await pg.schema.hasTable('players').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('players', (table) => {
          table.increments();
          table.uuid('uuid').notNullable().unique();
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table players');
          for (let i = 0; i < 10; i++) {
            const uuid = Helpers.generateUUID();
            await pg.table('players').insert({
              uuid
            })
          }
        }).catch((error) => {
          console.log(error);
        });;

    }
  });

  await pg.schema.hasTable('games').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('games', (table) => {
          table.increments();
          table.uuid('uuid').notNullable().unique();
          table.string('title');
          table.integer("rounds");
          table.integer("duration");
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table games');
        })
        .catch((error) => {
          console.log(error);
        });
    }
    
  });

  await pg.schema.hasTable('played_games').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('played_games', (table) => {
          table.increments();
          table.uuid('game_id').unsigned().references("uuid").inTable("games").onDelete("CASCADE").onUpdate("CASCADE");
          table.uuid('player_id').unsigned().references("uuid").inTable("players").onDelete("CASCADE").onUpdate("CASCADE");
          table.boolean("winner");
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table played_games');

        }).catch((error) => {
          console.log(error);
        });

    }
  });
}
initialiseTables();

module.exports = app;